File: backend/app/models/transcript.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel


# Define a Pydantic model for the segment structure
class TranscriptSegment(BaseModel):
    """
    TimestampSegment model representing a segment of audio with its metadata.
    Attributes:
        id (int): Unique identifier for the segment.
        start (float): Start time of the segment in seconds.
        end (float): End time of the segment in seconds.
        text (str): Transcribed text for the segment.
    """

    id: int
    start: float
    end: float
    text: str


class Transcript(BaseModel):
    """
    Transcript model representing a transcript with its metadata and content.

    Attributes:
        id (str): Unique identifier for the transcript.
        title (str): Title of the transcript.
        audio_path (Optional[str]): Path to the associated audio file.
        timestamps (List[TimestampSegment]): List of timestamp segments.
        unedited_id (str): ID of the initial transcript.
        previous_id (str): ID of the previous version.
        created_at (datetime): Timestamp when the transcript was created.
        updated_at (datetime): Timestamp when the transcript was last updated.
    """

    id: str  # Changed from int to string for content-based hash
    title: str
    audio_path: Optional[str] = None
    segments: List[TranscriptSegment] = []
    unedited_id: str  # ID of the initial transcript
    previous_id: str  # ID of the previous version
    created_at: datetime
    updated_at: datetime


class Config:
    from_attributes = True

File: src/models/transcript.ts

/**
 * Represents single segment of a transcript.
 * 
 * @interface TranscriptSegment
 * @property {string} s_id - Unique identifier for the transcript (s_id to avoid conflicts).
 * @property {number} start - Start time of the segment in seconds.
 * @property {number} end - End time of the segment in seconds.
 * @property {string} text - Text content of the segment.
 * @property {string} speaker - Speaker identifier (optional).
 */

export interface TranscriptSegment  {
  s_id: number;
  speaker?: string;
  start: number;
  end: number;
  text: string;
}

/**
 * Represents a transcript object for use with the transcript editor
 * 
 * @interface Transcript
 * @property {string} id - Unique identifier for the transcript.
 * @property {string} title - Title of the transcript.
 * @property {string} data_id - Identifier for this transcript data.
 * @property {string} previous_did - Identifier for previous transcript data.
 * @property {string} file_id - Path to the audio file or null if no audio available.
 * @property {TranscriptSegment[]} segments - Array of transcript segments.
 * @property {string} updated_at - ISO timestamp of when the transcript was last updated.
 * @property {string} created_at - ISO timestamp of when the transcript was created.
 */
export interface Transcript {
  id: string;
  title: string;
  data_id: string;
  previous_did: string;
  file_id: string;
  file_url: string;
  created_at: string;
  updated_at: string;
  segments: TranscriptSegment[];
}


/**
 * Represents a transcript metadata for use with the sidebar
 * 
 * @interface TranscriptMeta
 * @property {string} id - Unique identifier for the transcript.
 * @property {string} title - Title of the transcript.
 * @property {string} updated_at - ISO timestamp of when the transcript was last updated.
 * @property {string} created_at - ISO timestamp of when the transcript was created.
 * @property {string} file_id - Path to the audio file or null if no audio available.
 * @property {data_type} file_type - Type of the transcript file.
 * @property {string} data_id - Identifier for this transcript data.
 */

export interface TranscriptMeta {
  id: string;
  title: string;
  created_at: string;
  updated_at: string;
  file_id: string;
  file_url: string;
  file_type: file_type;
  data_id: string;
}

export type file_type = "audio" | "none"
export const NO_FILE = "no_file"

